### PR TITLE
[SPARK-9192][SQL] add initialization phase for nondeterministic expression

### PR DESCRIPTION
**Pull request by cloud-fan**
Currently nondeterministic expression is broken without a explicit initialization phase.

Let me take `MonotonicallyIncreasingID` as an example. This expression need a mutable state to remember how many times it has been evaluated, so we use `@transient var count: Long` there. By being transient, the `count` will be reset to 0 and **only** to 0 when serialize and deserialize it, as deserialize transient variable will result to default value. There is _no way_ to use another initial value for `count`, until we add the explicit initialization phase.

Another use case is local execution for `LocalRelation`, there is no serialize and deserialize phase and thus we can't reset mutable states for it.

### CODE REVIEW

**Comment by rxin** [comment]
I discussed with @marmbrus on this. Here's the suggestion:

1. Get rid of Initializable trait, and move initialize into Expression itself, mark it final. Its implementation recursively initialize children, and then calls an internal version for the expression to initialize itself. Mark an initialized flag as true after that.

2. Add a check in various implementations of nullSafeEval to fail if initialized is false; this helps us catch problems.

3. A few places we are using Cast expression to do proper casting. Consider moving those out of cast into a cast util class, and call those directly instead since they should be stateless. (Are they always stateless? Are timestamp/etc rely on Calendar?)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -38,99 +38,105 @@ trait CheckAnalysis {
...
+      case e: Generator => e
+    }).length > 1
   }
 
   def checkAnalysis(plan: LogicalPlan): Unit = {
     // We transform up and order the rules so as to catch the first possible failure instead
     // of the result of cascading resolution failures.
-    plan.foreachUp {
-
-      case operator: LogicalPlan =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: `case operator: LogicalPlan` catches all cases, so why not just use a normal function here instead of pattern match?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -38,99 +38,105 @@ trait CheckAnalysis {
...
+               |Failure when resolving conflicting references in Join:
+               |$plan
+                |Conflicting attributes: ${conflictingAttributes.mkString(",")}
+                |""".stripMargin)
+
+        case o if !o.resolved =>
+          failAnalysis(
+            s"unresolved operator ${operator.simpleString}")
+
+        case o if o.expressions.exists(!_.deterministic) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: here is the newly added checking rule for nondeterministic expressions.

**Comment by cloud-fan** [comment]
hi @marmbrus @rxin , I have updated the description and code, can you review it for me?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -38,10 +37,10 @@ trait CheckAnalysis {
     throw new AnalysisException(msg)
   }
 
-  def containsMultipleGenerators(exprs: Seq[Expression]): Boolean = {
+  protected def containsMultipleGenerators(exprs: Seq[Expression]): Boolean = {
     exprs.flatMap(_.collect {
-      case e: Generator => true
-    }).nonEmpty
+      case e: Generator => e
+    }).length > 1
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: a un-related but small fix: check multiple should use `length > 1` instead of `nonEmpty`

**Comment by cloud-fan** [comment]
cc @rxin @marmbrus It's ready for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: oops

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 4]
```diff
@@ -129,6 +128,14 @@ trait CheckAnalysis {
             failAnalysis(
               s"unresolved operator ${operator.simpleString}")
 
+          case o if o.expressions.exists(!_.deterministic) &&
+            !o.isInstanceOf[Project] && !o.isInstanceOf[Filter] =>
+            failAnalysis(
+              s"""nondeterministic expressions are only allowed in Project or Filter, found:
+                 | ${o.expressions.map(_.prettyString).mkString(",")}
+                  |in operator ${operator.simpleString}
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: this is not aligned ...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 5]
```diff
@@ -129,6 +128,14 @@ trait CheckAnalysis {
             failAnalysis(
               s"unresolved operator ${operator.simpleString}")
 
+          case o if o.expressions.exists(!_.deterministic) &&
+            !o.isInstanceOf[Project] && !o.isInstanceOf[Filter] =>
+            failAnalysis(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: you make it seem like a user error here, but if we get here, we have a bug in the system.

maybe we should just add throw illegalstateexception

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -910,6 +913,30 @@ class Analyzer(
         Project(finalProjectList, withWindow)
     }
   }
+
+  /**
+   * Pulls out nondeterministic expressions from unary LogicalPlan which is not Project or Filter,
+   * put them into an inner Project and finally project them away at the outer Project.
+   */
+  object PullOutNondeterministic extends Rule[LogicalPlan] {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: can you add a unit test for this suite?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: I'd like to do it in another PR, as here in not the only place that need to update in CheckAnalysis.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -910,6 +913,34 @@ class Analyzer(
...
+    override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+      case p: Project => p
+      case f: Filter => f
+
+      // todo: It's hard to write a general rule to pull out nondeterministic expressions
+      // from LogicalPlan, currently we only do it for UnaryNode which has same output
+      // schema with its child.
+      case p: UnaryNode if p.output == p.child.output && p.expressions.exists(!_.deterministic) =>
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: cc @marmbrus  / @yhuai

**Comment by rxin** [comment]
I'm going to merge this one first. Would be great to get a holistic review before 1.5 release on all the random related optimizer/analyzer changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 5 above)
Comment (jeanlyn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Hi, @cloud-fan .Can it support for join operation? Sometimes we can use some `non deterministic ` expression to eval some pointless join keys(with respect to business logic) avoiding data skew.
For example
```sql
SELECT src.key, src.value, src1.value 
FROM src 
JOIN src1
ON UPPER((CASE WHEN (src.key IS NULL OR src.key = '' ) THEN CAST( (-RAND() * 10000000 ) AS string ) ELSE src.key END )) = UPPER(src1.key)
```
What do you think?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: We have `Repartition` operator to do this job, maybe you can try that instead of doing it manually?

Comment (jeanlyn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Yep, we need to change this sql manually after this behavior.
